src/userobjects/MDRunBase.C

```diff
@@ -0,0 +1,124 @@
+/*                                                                    */
+/*            Copyright 2017 Battelle Energy Alliance, LLC            */
+/*                        ALL RIGHTS RESERVED                         */
+/**********************************************************************/
+
+#include "MDRunBase.h"
+#include "MooseMesh.h"
+
+#include "libmesh/mesh_tools.h"
+
+// custom data load and data store methods for a class with virtual members (vtable pointer must not
+// be (un)serialized)
+template <>
+inline void
+dataStore(std::ostream & stream, MDRunBase::MDParticles & pl, void * context)
+{
+  dataStore(stream, pl.pos, context);
+  dataStore(stream, pl.pos_old, context);
+  dataStore(stream, pl.vel, context);
+  dataStore(stream, pl.vel_old, context);
+  dataStore(stream, pl.id, context);
+  dataStore(stream, pl.elem_id, context);
+}
+
+template <>
+inline void
+dataLoad(std::istream & stream, MDRunBase::MDParticles & pl, void * context)
+{
+  dataLoad(stream, pl.pos, context);
+  dataLoad(stream, pl.pos_old, context);
+  dataLoad(stream, pl.vel, context);
+  dataLoad(stream, pl.vel_old, context);
+  dataLoad(stream, pl.id, context);
+  dataLoad(stream, pl.elem_id, context);
+}
+
+template <>
+InputParameters
+validParams<MDRunBase>()
+{
+  InputParameters params = validParams<GeneralUserObject>();
+  params.set<ExecFlagEnum>("execute_on") = EXEC_TIMESTEP_BEGIN;
+  params.suppressParameter<ExecFlagEnum>("execute_on");
+
+  return params;
+}
+
+MDRunBase::MDRunBase(const InputParameters & parameters)
+  : GeneralUserObject(parameters),
+    _mesh(_subproblem.mesh()),
+    _dim(_mesh.dimension()),
+    _nproc(_app.n_processors()),
+    _bbox(_nproc)
+{
+}
+
+void
+MDRunBase::initialSetup()
+{
+  for (unsigned int j = 0; j < _nproc; ++j)
+    _bbox[j] = MeshTools::create_processor_bounding_box(_mesh, j);
+}
+
+void
+MDRunBase::timestepSetup()
+{
+  // update/init subdomain bounding boxes
+  for (unsigned int j = 0; j < _nproc; ++j)
+    _bbox[j] = MeshTools::create_processor_bounding_box(_mesh, j);
+
+  // callback for updating md particle list
+  updateParticleList();
+}
+
+void
+MDRunBase::updateKDTree()
+{
+  _kd_tree = libmesh_make_unique<KDTree>(_md_particles.pos, 50);
+}
+
+void
+MDRunBase::mapMDParticles()
+{
+  // clear data structures
+  _elem_particles.clear();
+  _md_particles.elem_id.assign(_md_particles.pos.size(), 0);
+
+  // loop over elements to check against bounding box
+  ConstElemRange & elem_range = *_mesh.getActiveLocalElementRange();
+  ConstElemRange::const_iterator el = elem_range.begin();
+  for (el = elem_range.begin(); el != elem_range.end(); ++el)
+  {
+    const Elem * elem = *el;
+
+    // find all points within an inflated bounding box
+    std::vector<std::pair<std::size_t, Real>> indices_dist;
+    BoundingBox bbox = elem->loose_bounding_box();
+    Point center = 0.5 * (bbox.min() + bbox.max());
+    Real radius = (bbox.max() - center).norm();
```

Comment:
Why not use centroid and h_max here?

Reply:
I need to make sure to construct a sphere that contains the element. `h_max` is the spacing between nodes and I am not sure that `h_max / 2` is equal or larger than the radius of the circumscribing sphere.

Reply:
Yes, you're right. I may have made this mistake already somewhere in the code... :-/

Reply:
The alternative is to compute the circumscribing sphere in `Elem`.  It requires computation of the centroid though and that might be expensive. I could push a corresponding function in libmesh.